File: JavaScript/7js 1/compose.js
/*
|--------------------------------------------------------------------------
| Composing squared strings
|--------------------------------------------------------------------------
|
*/

function compose(s1, s2) {
   let x1 = s1.split("\n");
   let x2 = s2.split("\n");
   let s = x1.map((x1, i) => x1.substr(0, i + 1) + x2[x2.length - i - 1].substr(0, x2.length - i));
   return s.join("\n");
}


function compose(s1, s2) {
   return s1
      .split("\n")
      .map(
         (x, i, arr, arr2 = s2.split("\n").reverse()) =>
            x.slice(0, i + 1) + arr2[i].slice(0, arr2.length - i)
      )
      .join("\n");
}


function compose(s1, s2, idx = 0) {
   return s1.replace(
      /\S+/g,
      (val) =>
         val.slice(0, idx + 1) +
         s2
            .split(`\n`)
            .reverse()
            [idx].slice(0, val.length - idx++)
   );
}


console.log(
   compose("byGt\nhTts\nRTFF\nCnnI", "jIRl\nViBu\nrWOb\nNkTB"),
   "bNkTB\nhTrWO\nRTFVi\nCnnIj"
);

File: Python/7py 1/compose.py
"""
|--------------------------------------------------------------------------
| Composing squared strings
|--------------------------------------------------------------------------
|
"""

def compose(s1, s2):
   x1, x2 = s1.split("\n"), s2.split("\n")
   res = [x1[i][:i + 1] + x2[-i - 1][:len(x2) - i] for i in range(len(x1))]
   return "\n".join(res)


def compose(s1, s2):
   return '\n'.join(e[0][:i+1]+e[1][:len(e[1])-i]for i,e in enumerate(zip(s1.split(),s2.split()[::-1])))


def compose(s1: str, s2: str) -> str:
   s1 = s1.split("\n")
   s2 = s2.split("\n")
   new_s = []
   for i, v in enumerate(s1):
      new_s.append(s1[i][:i+1] + s2[-i-1][:len(s2) - i]) 
   return '\n'.join(new_s)


print(
   compose("byGt\nhTts\nRTFF\nCnnI", "jIRl\nViBu\nrWOb\nNkTB"),
   "bNkTB\nhTrWO\nRTFVi\nCnnIj"
);